File: sgi-backend/app/schemas/usuario.py
from pydantic import BaseModel, EmailStr
from typing import Optional, List
from datetime import datetime


# =========================================
# REQUEST SCHEMAS
# =========================================

class UsuarioCreate(BaseModel):
    empleado_id: int
    correo_corp: EmailStr
    password: str
    roles: List[int] = []


class UsuarioUpdate(BaseModel):
    correo_corp: Optional[EmailStr] = None
    is_active: Optional[bool] = None
    debe_cambiar_pass: Optional[bool] = None
    is_bloqueado: Optional[bool] = None
    roles: Optional[List[int]] = None


class UsuarioChangePassword(BaseModel):
    password: str


class UsuarioAssignRoles(BaseModel):
    roles: List[int]


# =========================================
# RESPONSE SCHEMAS
# =========================================

class RolDropdown(BaseModel):
    id: int
    nombre: str
    descripcion: Optional[str] = None


class EmpleadoSinUsuario(BaseModel):
    id: int
    nombre_completo: str
    nro_documento: str
    area_nombre: Optional[str] = None
    cargo_nombre: Optional[str] = None


class UsuarioResponse(BaseModel):
    id: int
    empleado_id: Optional[int] = None
    empleado_nombre: Optional[str] = None
    correo_corp: str
    is_active: bool
    is_bloqueado: bool
    ultimo_acceso: Optional[datetime] = None
    debe_cambiar_pass: bool = False
    roles: Optional[str] = None
    total_registros: Optional[int] = None


class UsuarioDetailResponse(BaseModel):
    id: int
    empleado_id: Optional[int] = None
    empleado_nombre: Optional[str] = None
    correo_corp: str
    is_active: bool
    is_bloqueado: bool
    ultimo_acceso: Optional[datetime] = None
    debe_cambiar_pass: bool
    intentos_fallidos: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    roles: List[RolDropdown] = []


class UsuarioPaginatedResponse(BaseModel):
    total: int
    page: int
    page_size: int
    total_pages: int
    data: List[UsuarioResponse]

File: sgi-frontend/src/types/usuario.ts
// Tipos para Usuarios

export interface Usuario {
    id: number;
    empleado_id: number | null;
    empleado_nombre: string | null;
    correo_corp: string;
    is_active: boolean;
    is_bloqueado: boolean;
    ultimo_acceso: string | null;
    debe_cambiar_pass: boolean;
    roles: string | null;
    total_registros?: number;
}

export interface UsuarioPaginationResponse {
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
    data: Usuario[];
}

export interface UsuarioOperationResult {
    success: number;
    message: string;
    id?: number;
}

export interface RolOption {
    id: number;
    nombre: string;
    descripcion?: string | null;
}

export interface EmpleadoSinUsuario {
    id: number;
    nombre_completo: string;
    nro_documento: string;
    area_nombre?: string | null;
    cargo_nombre?: string | null;
}

export interface UsuarioCreate {
    empleado_id: number;
    correo_corp: string;
    password: string;
    roles: number[];
}

export interface UsuarioUpdate {
    correo_corp?: string | null;
    is_active?: boolean | null;
    debe_cambiar_pass?: boolean | null;
    is_bloqueado?: boolean | null;
    roles?: number[] | null;
}

export interface UsuarioDetail extends Usuario {
    intentos_fallidos: number;
    created_at: string | null;
    updated_at: string | null;
    roles_list: RolOption[];
}
